src/frontend-common/save_state_selector_ui.cpp

```diff
@@ -187,52 +187,120 @@ void SaveStateSelectorUI::Draw()
+          const ImVec2 p_end(p_start.x + window_width, p_start.y + item_height);
+          ImGui::GetWindowDrawList()->AddRectFilled(p_start, p_end, ImColor(0.22f, 0.30f, 0.34f, 0.9f), rounding);
+        }
+
+        if (entry.preview_texture)
+        {
+          ImGui::SetCursorPosY(y_start + padding);
+          ImGui::SetCursorPosX(padding);
+          ImGui::Image(reinterpret_cast<ImTextureID>(entry.preview_texture->GetHandle()), image_size);
+        }
 
-        const ImVec2 p_start(ImGui::GetCursorScreenPos());
-        const ImVec2 p_end(p_start.x + window_width, p_start.y + item_height);
-        ImGui::GetWindowDrawList()->AddRectFilled(p_start, p_end, ImColor(0.22f, 0.30f, 0.34f, 0.9f), rounding);
-      }
-
-      if (entry.preview_texture)
-      {
         ImGui::SetCursorPosY(y_start + padding);
-        ImGui::SetCursorPosX(padding);
-        ImGui::Image(reinterpret_cast<ImTextureID>(entry.preview_texture->GetHandle()), image_size);
-      }
-
-      ImGui::SetCursorPosY(y_start + padding);
-
-      ImGui::Indent(text_indent);
 
-      ImGui::Text("%s Slot %d", entry.global ? "Global" : (entry.game_code.empty() ? "Game" : entry.game_code.c_str()),
-                  entry.slot);
-      ImGui::TextUnformatted(entry.title.c_str());
-      ImGui::TextUnformatted(entry.formatted_timestamp.c_str());
-      ImGui::TextUnformatted(entry.path.c_str());
+        ImGui::Indent(text_indent);
+
+        if (entry.global)
+        {
+          ImGui::Text(m_host_interface->TranslateString("SaveStateSelectorUI", "Global Slot %d"), entry.slot);
+        }
+        else if (entry.game_code.empty())
+        {
+          ImGui::Text(m_host_interface->TranslateString("SaveStateSelectorUI", "Gane Slot %d"), entry.slot);
+        }
+        else
+        {
+          ImGui::Text(m_host_interface->TranslateString("SaveStateSelectorUI", "%s Slot %d"), entry.game_code.c_str(),
+                      entry.slot);
+        }
+        ImGui::TextUnformatted(entry.title.c_str());
+        ImGui::TextUnformatted(entry.formatted_timestamp.c_str());
+        ImGui::TextUnformatted(entry.path.c_str());
+
+        ImGui::Unindent(text_indent);
+      }
+    }
+    ImGui::EndChild();
 
-      ImGui::Unindent(text_indent);
+    ImGui::BeginChild("##legend", ImVec2(0, 0), false,
+                      ImGuiWindowFlags_NoFocusOnAppearing | ImGuiWindowFlags_NoInputs | ImGuiWindowFlags_NoTitleBar |
+                        ImGuiWindowFlags_NoScrollbar);
+    {
+      ImGui::SetCursorPosX(padding);
+      ImGui::BeginTable("table", 2);
+
+      auto strip_device_name = [](std::string str) {
+        std::string result;
+
+        auto slash_pos = str.find_first_of('/');
+        if (slash_pos != str.npos)
+        {
+          result = str.substr(slash_pos + 1);
+        }
+        else
+        {
+          result = std::move(str);
+        }
+        return result;
+      };
+
+      const std::string load_savestate_hotkey =
```

Comment:
Not really a fan of having these lookups - I try to avoid per-frame heap allocations. Could we cache/stash them somewhere in the class?

Reply:
Sure, but then we need to subscribe to a signal if those ever change, I think :(